File: go/value.go
// Package haltrule is the Go port of the haltrule spec: the policy of stopping
// well. Pure policy - no I/O, no timers, no clock, no randomness - and the
// standard library only. ../fixtures is the contract every port must satisfy;
// ../spec/README.md is the spec in words. Behaviour, not names, must match the
// other ports: Go identifiers follow Go's conventions.
//
// What the other ports check at run time, this one mostly cannot be handed at
// all: Value below is closed, a cap is an int64, and an absent argument is a
// nil pointer. That is the same contract, kept by the type instead of by code.
package haltrule

// Value is the spec's value model: null, booleans, numbers, strings, lists,
// and maps with string keys. It is closed - isValue is unexported, so no type
// outside this package can be a Value - which makes the model a type rather
// than a check. A nil Value is null.
//
// Float is here although the model holds only integers: a caller must be able
// to hand over the fractional number that a digest halts on. A number is an
// integer when its value is integral, however it is held, so Float(2) and
// Int(2) canonicalize alike.
type Value interface{ isValue() }

type (
	Bool   bool
	Int    int64
	Float  float64
	String string
	List   []Value
	Map    map[string]Value
)

func (Bool) isValue()   {}
func (Int) isValue()    {}
func (Float) isValue()  {}
func (String) isValue() {}
func (List) isValue()   {}
func (Map) isValue()    {}

File: rust/haltrule/src/value.rs
use alloc::collections::BTreeMap;
use alloc::string::String;
use alloc::vec::Vec;

/// The spec's value model: null, booleans, numbers, strings, lists, and maps
/// with string keys. An enum is closed, so a value outside the model cannot
/// be built - the same contract the other ports keep with a check.
///
/// Float is here although the model holds only integers: a caller must be
/// able to hand over the fractional number that a digest halts on. A number
/// is an integer when its value is integral, however it is held, so
/// `Float(2.0)` and `Int(2)` canonicalize alike.
#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Null,
    Bool(bool),
    Int(i64),
    Float(f64),
    Text(String),
    List(Vec<Value>),
    Map(Map),
}

/// A map of the value model. Its order is the keys' own; the canonical form
/// and a result line both sort by UTF-16 code unit, which is not this order.
pub type Map = BTreeMap<String, Value>;

/// An argument outside a part's contract. The call fails and changes nothing:
/// a refusal is not a verdict and carries no reason. The text is for a person
/// reading a stack trace and is not part of conformance.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Refused(pub &'static str);
